src/lib/FreeC/Pass/EtaConversionPass.hs

```diff
@@ -78,18 +135,91 @@ etaConversionPass ast = do
 -- Function declarations                                                     --
 -------------------------------------------------------------------------------
 
--- | Applies 'etaConvertExpr' to the right-hand side of the given function
---   declaration.
+-- | Depending on the presence or absence of missing top-level arguments, applies 'modifyTopLevel' or 'etaConvertExpr' to the right-hand side of the given function declaration to ensure all functions 
+-- and constructors on the right-hand side are fully applied. 
+-- The missing top-level arguments are also added to the left-hand
+-- side of the declaration and the function's type and the environment
+-- are updated accordingly. 
 etaConvertFuncDecl :: IR.FuncDecl -> Converter IR.FuncDecl
 etaConvertFuncDecl funcDecl = do
-  rhs' <- etaConvertExpr (IR.funcDeclRhs funcDecl)
-  return funcDecl { IR.funcDeclRhs = rhs' }
+  let rhs = IR.funcDeclRhs funcDecl
+  newArgNumber <- findMinMissingArguments rhs
+  -- Only perform top-level eta conversion when all alternatives for right-hand sides are missing at least one argument.
+  if newArgNumber == 0
+    then do
+      rhs' <- etaConvertExpr rhs
+      return funcDecl { IR.funcDeclRhs = rhs' }
```

Comment:
Shouldn't the `else` branch cover this case already?

Reply:
It does, but that will cause the function to needlessly descend into the function's right-hand side again in `etaConvertTopLevel` and also needlessly 'update' the environment, and a lot of functions will also be called only to be cut short immediately because they were called with an empty list. 
This is really just a shortcut to avoid all of that.